Source/JavaScriptCore/assembler/MacroAssemblerX86_64.h

```diff
@@ -3375,6 +3369,75 @@ class MacroAssemblerX86_64 : public MacroAssemblerX86Common {
         }
     }
 
+    void vectorUshl8(FPRegisterID input, FPRegisterID shift, FPRegisterID dest, FPRegisterID tmp1, FPRegisterID tmp2)
+    {
+        RELEASE_ASSERT(supportsAVX());
+
+        // Unpack and zero-extend low input bytes.
+        m_assembler.vxorps_rrr(tmp2, tmp2, tmp2);
+        m_assembler.vpunpcklbw_rrr(input, tmp2, tmp1);
+
+        // Word-wise shift low input bytes into tmp1.
+        m_assembler.vpsllw_rrr(shift, tmp1, tmp1);
+
+        // Unpack and zero-extend high input bytes.
+        m_assembler.vpunpckhbw_rrr(input, tmp2, tmp2);
+
+        // Word-wise shift high input bytes into tmp2.
+        m_assembler.vpsllw_rrr(shift, tmp2, tmp2);
+
+        // Mask away higher bits of left-shifted results.
+        m_assembler.vpcmpeqw_rrr(dest, dest, dest);
```

Comment:
You are early-def'ing dest here, but the air opcode only claims to def it

Reply:
Can scratches interfere with defs? My assumption was that scratch registers would be allocated separately from the other parameters, in which case this should still be fine (since we only write `dest` after we finish using `input` and `shift`). Good catch, I'll check what we do in the register allocator, I guess worst comes to worst we can make `dest` a use/def.

Reply:
It seems to be the case that scratches are always allocated separately from defs within the same instruction, so this code should work. But, in trying to avoid writing to dest, I ended up finding a different instruction sequence that is probably a bit faster and still passes the tests, so I'll be amending this PR to use that.